File: insertionSortList.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # @param head, a ListNode
    # @return a ListNode
    def insertionSortList(self, head):
    	if head is None:
    		return head
    	before = None
        toSort = head
        next = toSort.next
        while toSort:
        	p = head
        	last = None
        	while p != toSort and p != next:
        		if p.val > toSort.val:
        			if last:
        				last.next = toSort
        			else:
        				head = toSort
        			if before:
        				before.next = next
        			toSort.next = p
        			break
        		last = p
        		p = p.next
        	if p == toSort:
        		before = toSort
        	toSort = next
        	if next:
        	   	next = next.next
        	# h = head
        	# while h:
        	# 	print h.val,
        	# 	h = h.next
        	# print 
        return head

s = Solution()
h = None
tail = None
for i in [-1,4,-3,2,1,5,6] * 200:
	if not h:
		h = ListNode(i)
		tail = h
	else:
		tail.next = ListNode(i)
		tail = tail.next
h = s.insertionSortList(h)
while h:
	print h.val,
	h = h.next

File: JAVA/insertionSortList.java
// Solution 1

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode insertionSortList(ListNode head) {
        if (head == null || head.next == null)
            return head;
        ListNode last = head, cur=head.next, next, temp, p, q;
        while(cur != null){
            next = cur.next;
            if (head.val > cur.val){
                cur.next = head;
                last.next = next;
                head = cur;
                cur = next;
                continue;
            }
            
            q = null;
            temp = head;
            while(temp != last && temp.val <= cur.val){
                q = temp;
                temp = temp.next;
            }
            if (temp.val <= cur.val){ // important, cannot change to temp != last, since they can be both unsatisfied, 4213 -> 1243
                last = cur;
                cur = next;
                continue;
            }
            else{
                p = q.next;
                q.next = cur;
                cur.next = p;
                last.next = next;
                cur = next;
            }
        }
        return head;
    }
}

// Solution 2

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode insertionSortList(ListNode head) {
        if (head == null || head.next == null)
            return head;
        ListNode last = head, cur=head.next, next, temp, p, q;
        while(cur != null){
            next = cur.next;
            if (head.val > cur.val){
                cur.next = head;
                last.next = next;
                head = cur;
                cur = next;
                continue;
            }
            if (last.val <= cur.val){
                last = cur;
                cur = next;
                continue;
            }
            q = null;
            temp = head;
            while(temp != last && temp.val <= cur.val){
                q = temp;
                temp = temp.next;
            }
            p = q.next;
            q.next = cur;
            cur.next = p;
            last.next = next;
            cur = next;
        }
        return head;
    }
}